pkg/noderesourcetopology/config.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package noderesourcetopology
+
+import (
+	"k8s.io/klog/v2"
+
+	topologyv1alpha2 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha2"
+)
+
+const (
+	AttributeScope  = "topologyManagerScope"
+	AttributePolicy = "topologyManagerPolicy"
+)
+
+// TODO: can we reuse constants from k/k/pkg/kubelet/cm/topology_manager ?
+const (
+	ScopeContainer = "container"
+	ScopePod       = "pod"
+)
+
+// TODO: can we reuse constants from k/k/pkg/kubelet/cm/topology_manager ?
+const (
+	PolicyNone           = "none"
+	PolicyBestEffort     = "best-effort"
+	PolicyRestricted     = "restricted"
+	PolicySingleNUMANode = "single-numa-node"
+)
+
+func IsValidScope(scope string) bool {
+	if scope == ScopeContainer || scope == ScopePod {
+		return true
+	}
+	return false
+}
+
+func IsValidPolicy(policy string) bool {
+	if policy == PolicyNone || policy == PolicyBestEffort || policy == PolicyRestricted || policy == PolicySingleNUMANode {
+		return true
+	}
+	return false
+}
+
+type TopologyManagerConfig struct {
+	Scope  string
+	Policy string
+}
+
+func (conf TopologyManagerConfig) IsComplete() bool {
```

Comment:
I have a question, it is valid not specify both `scope` and `policy` in Kubelet config, and their will be defaulted to `container` for scope and `none` for policy. Do we assume that NFD(or whichever component will advertise those config option) will make sure to make explicit conversion from "" to default value or we should just assume here that those values weren't set and we can just change "" to  default values here.

Reply:
that's a good point. I don't have an answer yet for NFD, and I'm revisiting this function and how we use it.

Reply:
These details should be discussed and documented, because they are a key part of the producer-consumer contract. I started a change (in this PR) to add this documentation in `pkg/noderesourcetopology/README.md`, let's continue here. If we find we need more space and discussion, we can move later to another forum (e.g. gh issue, google doc...)

Reply:
Yeah, I just feel producer should not do a lot of preprocessing so we can just emit what's configured in kubelet and not bother about complicated logic. IMHO setting defaults for those values should happen in consumer(in this case NodeResourceTopologyMatch plugin).

Reply:
> Yeah, I just feel producer should not do a lot of preprocessing so we can just emit what's configured in kubelet and not bother about complicated logic. IMHO setting defaults for those values should happen in consumer(in this case NodeResourceTopologyMatch plugin).

yes, I'm converging towards this point. What I think now (subject to change/further thinking):

- producers (NFD topology updater) should try their hardest to provide the values, but it the kubelet doesn't provide sufficient introspection mechanism, they should NOT guess or worse hardcode values. I think we're already in agreement here.
- consumer (scheduler plugin) CAN have defaults, because out of necessity we duplicate TM logic and constants anyway. But I'm still reluctant to hardcode defaults here, let's see.

Reply:
>     * consumer (scheduler plugin) CAN have defaults, because out of necessity we duplicate TM logic and constants anyway. But I'm still reluctant to hardcode defaults here, let's see.

No biggie for us when it comes with empty `policy` NodeResourceTopologyMatch  is noop for `none` policy. But we should probably assume that if `scope` is empty it means it is `container` scope(I bet many Kubernetes configurations uses TopologyManager but do not set `scope` and just rely on default here)

Reply:
I'm not fully sold yet, but hardcoding default policy=none and default scope=container and using them as fallback if nothing is given from the producer (NFD) is something we can surely try out and evaluate.

Reply:
Added in the last upload as promised